File: UserSpace/Cpp/AxionMass/AxionMass.cpp
#include <iostream>
#include <iomanip> 
#include <cmath> 
#include <string> 

#include"MiMeS.hpp"


#ifndef LONG
    #define LONG 
#endif

#ifndef LD
    #define LD LONG double
#endif

int main(int argc, char **argv){ 

    if(argc!=4){
        std::cerr<<"usage: AxionMass.run fa minT maxT\n";
        std::cerr<<"With:\n \
        fa: PQ scale.\n \
        minT: minimum interpolation tempareture.\n \
        maxT: maximum interpolation tempareture.\n\n \
        Beyond these limits, everything is taken constant, so one can use minT=1e-5 GeV and maxT=1e3 with good accuracy.\n";

        return 1;
    }

    LD fa=atof(argv[1]);

    LD minT=atof(argv[2]);
    LD maxT=atof(argv[3]);
    


    mimes::AxionMass<LD> axionMass(chi_PATH,minT,maxT);
    /*set ma2 for T>TMax*/
    LD TMax=axionMass.getTMax();    
    LD chiMax=axionMass.getChiMax();    
    axionMass.set_ma2_MAX( [&chiMax,&TMax](LD T, LD fa){return chiMax/fa/fa*std::pow(T/TMax,-8.16);});  
    
    /*set ma2 for T<TMin*/
    LD TMin=axionMass.getTMin();  
    LD chiMin=axionMass.getChiMin();    
    axionMass.set_ma2_MIN( [&chiMin,&TMin](LD T, LD fa){return chiMin/fa/fa;});


    if(minT==maxT){
        std::cout<<axionMass.ma2(minT,fa)<<"\t"<<axionMass.ma2(maxT,fa)<<"\n";
        return 0;
    }
    // print N points
    unsigned int N=50;
    
    // take logarithmically spaced points 
    std::vector<LD> T;
    /*---use logspace like this:---*/
    // mimes::util::logspace<LD>(std::log10(minT), std::log10(maxT),N,T);
    
    /*---or like this:---*/
    T = mimes::util::logspace<LD>(std::log10(minT), std::log10(maxT),N);
    
    std::cout<<"-----Interpolation-----\n";
    std::cout<<"T[GeV]\tm_a^2[GeV^2]\n";
    for(unsigned int i = 0; i < N; ++i){
        std::cout<<T[i]<<"\t"<<axionMass.ma2(T[i],fa)<<"\n";
    }
    

    return 0;
}

File: UserSpace/Python/AxionMass.py
from numpy import logspace, pi


from sys import path as sysPath
from sys import path as sysPath
sysPath.append('../../src')

#load the module
from interfacePy.AxionMass import AxionMass


axionMass = AxionMass(r'../../src/data/chi.dat',0,1e5)

# this is the axion mass squared beyond the interpolation limits for the current data 
TMax=axionMass.getTMax() 
chiMax=axionMass.getChiMax()
TMin=axionMass.getTMin() 
chiMin=axionMass.getChiMin()

axionMass.set_ma2_MAX( lambda T,fa: chiMax/fa/fa*pow(T/TMax,-8.16) )
axionMass.set_ma2_MIN( lambda T,fa: chiMin/fa/fa )


def ma2(T,fa):
    TQCD=150*1e-3;
    ma20=3.1575e-05/fa/fa;
    if T<=TQCD:
        return ma20;
    return ma20*pow((TQCD/T),8.16)
axionMassFunc = AxionMass(ma2)


print(axionMassFunc.ma2(3,1))
print(axionMass.ma2(3,1))

print(axionMass.ma2(0,1))
print(axionMassFunc.ma2(0,1))

del axionMass 
del axionMassFunc 